src/NHibernate/Impl/SessionImpl.cs

```diff
@@ -2221,7 +2221,10 @@ public ISession GetChildSession()
 
 				CheckAndUpdateSessionStatus();
 
-				if (_childSession == null)
+				var childImpl = _childSession as SessionImpl;
+
+				// if child session never existed or has already been disposed, create new
+				if (childImpl == null || childImpl.IsAlreadyDisposed)
```

Comment:
There are 2 points to implement the correct logic here:

1. It can be back-ported (why? because this is actually a bug) to previous versions without a lot of work required

2. I'm not yet sure that the improvement will make it to 5.0